htlcswitch/payment_result.go

```diff
@@ -297,3 +437,148 @@ func (store *networkResultStore) CleanStore(keep map[uint64]struct{}) error {
+//
+// NOTE: This function is NOT safe for concurrent access.
+func (store *networkResultStore) FetchPendingAttempts() ([]uint64, error) {
+	var pending []uint64
+	err := kvdb.View(store.backend, func(tx kvdb.RTx) error {
+		bucket := tx.ReadBucket(networkResultStoreBucketKey)
+		if bucket == nil {
+			return nil
+		}
+
+		return bucket.ForEach(func(k, v []byte) error {
+			// If the key is not 8 bytes, it's not a valid attempt
+			// ID.
+			if len(k) != 8 {
+				log.Warnf("Found invalid key of length %d in "+
+					"network result store", len(k))
+
+				return nil
+			}
+
+			// Deserialize the result to check its type.
+			r := bytes.NewReader(v)
+			result, err := deserializeNetworkResult(r)
+			if err != nil {
+				// If we can't deserialize, we'll log it and
+				// continue. The result will be removed by a
+				// call to CleanStore.
+				log.Warnf("Unable to deserialize result for "+
+					"key %x: %v", k, err)
+
+				return nil
+			}
+
+			// If the result is a pending result, add the attempt
+			// ID to our list.
+			if result.msg.MsgType() == pendingHtlcMsgType {
+				attemptID := binary.BigEndian.Uint64(k)
+				pending = append(pending, attemptID)
+			}
+
+			return nil
+		})
+	}, func() {
+		pending = nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pending, nil
+}
+
+// FailAttempt transitions an initialized attempt from PENDING to FAILED,
+// recording the provided reason. This is the synchronous rollback mechanism for
+// attempts that fail before being committed to the forwarding engine. It
+// returns an error if the underlying storage fails.
+//
+// NOTE: This method is specifically for attempts that have been initialized
+// via InitAttempt but fail *before* being dispatched to the network. Normal
+// failures (e.g., from an HTLC being failed on-chain or by a peer) are
+// recorded via the StoreResult method and should not use FailAttempt.
+func (store *networkResultStore) FailAttempt(attemptID uint64,
+	linkErr *LinkError) error {
+
+	// We get a mutex for this attempt ID to ensure consistency between the
+	// database state and the subscribers in case of concurrent calls.
+	store.attemptIDMtx.Lock(attemptID)
+	defer store.attemptIDMtx.Unlock(attemptID)
+
+	// First, verify that the attempt exists and is in the pending state,
+	// otherwise we cannot fail it.
+	err := store.verifyAttemptIsPending(attemptID)
+	if err != nil {
+		return fmt.Errorf("cannot fail attempt %d: %w", attemptID, err)
+	}
+
+	// The attempt to send the htlc failed before it was ever dispatched.
+	// We will write a failure result to the store to unblock any potential
+	// subscribers awaiting a final result.
+	failureResult, err := newInternalFailureResult(linkErr)
+	if err != nil {
+		return fmt.Errorf("failed to create failure message for "+
+			"attempt %d: %w", attemptID, err)
+	}
+
+	if err := store.storeResult(attemptID, failureResult); err != nil {
+		return err
+	}
+
+	store.notifySubscribers(attemptID, failureResult)
```

Comment:
We don't want to notify in case this was a pending attempt, right? Therefore it's also not needed to separate out `notifySubscribers`.

Reply:
Based on some of the tests I have in `switch_test.go`, this is what allows a client hanging on `GetAttemptResult` for an initialized but not actually dispatched HTLC attempt to unblock.

Currently the `GetAttemptResult` logic has been updated and will proceed to call `SubscribeResult` if it receives `ErrAttemptResultPending`. The transition from a `INIT` state to a `FAILED` state is what will notify such a caller that the attempt is actually failed so they can unblock and possibly retry via another attempt ID.